### PR TITLE
Make prepare engine step of recovery source non-blocking

### PR DESCRIPTION
**Pull request by dnhatn**
Relates #37174

### CODE REVIEW

File: server/src/main/java/org/elasticsearch/indices/recovery/PeerRecoveryTargetService.java [hunk 1]
```diff
@@ -431,13 +431,13 @@ public static long getStartingSeqNo(final Logger logger, final RecoveryTarget re
...
         @Override
-        public void messageReceived(RecoveryPrepareForTranslogOperationsRequest request, TransportChannel channel,
-                                    Task task) throws Exception {
-            try (RecoveryRef recoveryRef = onGoingRecoveries.getRecoverySafe(request.recoveryId(), request.shardId()
-            )) {
-                recoveryRef.target().prepareForTranslogOperations(request.isFileBasedRecovery(), request.totalTranslogOps());
+        public void messageReceived(RecoveryPrepareForTranslogOperationsRequest request, TransportChannel channel, Task task) {
+            try (RecoveryRef recoveryRef = onGoingRecoveries.getRecoverySafe(request.recoveryId(), request.shardId())) {
+                final ActionListener<TransportResponse> listener =
+                    new HandledTransportAction.ChannelActionListener<>(channel, Actions.PREPARE_TRANSLOG, request);
```
Comment (s1monw) [inline server/src/main/java/org/elasticsearch/indices/recovery/PeerRecoveryTargetService.java]: unrelated but I think we can move `ChannelActionListener` into it's own class

File: server/src/main/java/org/elasticsearch/indices/recovery/RecoverySourceHandler.java [hunk 2]
```diff
@@ -197,51 +197,48 @@ public void recoverToTarget(ActionListener<RecoveryResponse> listener) {
...
-                logger.trace("all operations up to [{}] completed, which will be used as an ending sequence number", endingSeqNo);
-                logger.trace("snapshot translog for recovery; current size is [{}]",
-                    shard.estimateNumberOfHistoryOperations("peer-recovery", startingSeqNo));
-            }
+                final long endingSeqNo = shard.seqNoStats().getMaxSeqNo();
+                /*
+                 * We need to wait for all operations up to the current max to complete, otherwise we can not guarantee that all
+                 * operations in the required range will be available for replaying from the translog of the source.
+                 */
+                cancellableThreads.execute(() -> shard.waitForOpsToComplete(endingSeqNo));
```
Comment (s1monw) [inline server/src/main/java/org/elasticsearch/indices/recovery/RecoverySourceHandler.java]: will / can we make this also non-blocking?

File: server/src/main/java/org/elasticsearch/indices/recovery/RecoverySourceHandler.java [hunk 3]
```diff
@@ -197,51 +197,48 @@ public void recoverToTarget(ActionListener<RecoveryResponse> listener) {
...
+                if (logger.isTraceEnabled()) {
+                    logger.trace("all operations up to [{}] completed, which will be used as an ending sequence number", endingSeqNo);
+                    logger.trace("snapshot translog for recovery; current size is [{}]",
+                        shard.estimateNumberOfHistoryOperations("peer-recovery", startingSeqNo));
+                }
+                final Translog.Snapshot phase2Snapshot = shard.getHistoryOperations("peer-recovery", startingSeqNo);
+                resources.add(phase2Snapshot);
+                // we can release the retention lock here because the snapshot itself will retain the required operations.
+                IOUtils.close(retentionLock);
```
Comment (s1monw) [inline server/src/main/java/org/elasticsearch/indices/recovery/RecoverySourceHandler.java]: can't you just call `retentionLock.close()`?

File: server/src/main/java/org/elasticsearch/indices/recovery/RecoverySourceHandler.java [hunk 4]
```diff
@@ -197,51 +197,48 @@ public void recoverToTarget(ActionListener<RecoveryResponse> listener) {
...
+                // we have to capture the max_seen_auto_id_timestamp and the max_seq_no_of_updates to make sure that these values
+                // are at least as high as the corresponding values on the primary when any of these operations were executed on it.
+                final long maxSeenAutoIdTimestamp = shard.getMaxSeenAutoIdTimestamp();
+                final long maxSeqNoOfUpdatesOrDeletes = shard.getMaxSeqNoOfUpdatesOrDeletes();
+                phase2(startingSeqNo, requiredSeqNoRangeStart, endingSeqNo, phase2Snapshot, maxSeenAutoIdTimestamp,
+                    maxSeqNoOfUpdatesOrDeletes, sendSnapshotStep);
+                sendSnapshotStep.whenComplete(
+                    r -> IOUtils.close(phase2Snapshot),
```
Comment (s1monw) [inline server/src/main/java/org/elasticsearch/indices/recovery/RecoverySourceHandler.java]: should we also close the `phase2Snapshot` in the error case?

File: server/src/main/java/org/elasticsearch/indices/recovery/RecoverySourceHandler.java [hunk 5]
```diff
@@ -484,16 +481,23 @@ public SendFileResult phase1(final IndexCommit snapshot, final Supplier<Integer>
...
         StopWatch stopWatch = new StopWatch().start();
-        logger.trace("recovery [phase1]: prepare remote engine for translog");
+        final ActionListener<Void> wrappedListener = ActionListener.wrap(
+            nullVal -> {
+                stopWatch.stop();
+                final TimeValue tookTime = stopWatch.totalTime();
+                logger.trace("recovery [phase1]: remote engine start took [{}]", tookTime);
+                listener.onResponse(tookTime);
+            },
+            e -> {
```
Comment (s1monw) [inline server/src/main/java/org/elasticsearch/indices/recovery/RecoverySourceHandler.java]: unnecessary `{}`?

**Review by s1monw — APPROVED** [review]
left some minors LGTM otherwise

File: server/src/main/java/org/elasticsearch/indices/recovery/RecoverySourceHandler.java (hunk 2 above)
Comment (dnhatn) [reply server/src/main/java/org/elasticsearch/indices/recovery/RecoverySourceHandler.java]: Good idea. I will do it.

**Comment by dnhatn** [comment]
Thanks @s1monw.

**Pull request merged**